app/utils/callback/lava.py

```diff
@@ -0,0 +1,234 @@
+            _get_lava_boot_meta(meta, step['metadata'])
+        elif step['name'] == 'bootloader-overlay':
+            _get_lava_bootloader_meta(meta, step['metadata'])
+
+def _add_boot_log(meta, log, base_path, errors):
+    """Parse and save kernel logs
+
+    Parse the kernel boot log in YAML format from a LAVA v2 job and save it as
+    plain text and HTML.
+
+    :param meta: The boot meta-data.
+    :type meta: dictionary
+    :param log: The kernel log in YAML format.
+    :type log: string
+    :param base_path: Path to the top-level directory where to store the files.
+    :type base_path: string
+    :param errors: Where to store the errors.
+    :type errors: dict
+    :return int The return code, the boot document id and errors.
+    """
+    ret_val = 201
+    log = yaml.load(log, Loader=yaml.CLoader)
+
+    dir_path = os.path.join(
+        base_path,
+        meta['job'],
+        meta['git_branch'],
+        meta['kernel'],
+        meta['arch'],
+        meta['defconfig_full'],
+        meta['lab_name'])
+    file_name = '-'.join(['boot', meta['board']])
+    files = tuple('.'.join([file_name, ext]) for ext in ['txt', 'html'])
+    meta['boot_log'], meta['boot_log_html'] = files
+    txt_path, html_path = (os.path.join(dir_path, f) for f in files)
+
+    try:
+        if not os.path.isdir(dir_path):
+            try:
+                os.makedirs(dir_path)
+            except OSError as e:
+                if e.errno != errno.EEXIST:
+                    raise e
+
+        with open(txt_path, 'w') as txt, open(html_path, 'w') as html:
+            utils.lava_log_parser.run(log, meta, txt, html)
+    except (OSError, IOError) as e:
+        msg = "Failed to save boot logs in {}".format(dir_path)
+        ret_val = 500
+        utils.LOG.exception(e)
+        utils.LOG.error(msg)
+        utils.errors.add_error(errors, ret_val, msg)
+
+    return ret_val
+
+def add_boot(job_data, lab_name, db_options, base_path=utils.BASE_PATH):
+    """Entry point to be used as an external task.
+
+    This function should only be called by Celery or other task managers.
+    Parse the boot data from a LAVA v2 job callback and save it along with
+    kernel logs.
+
+    :param job_data: The JSON data from the callback.
+    :type job_data: dict
+    :param lab_name: Name of the LAVA lab that posted the callback.
+    :type lab_name: string
+    :param db_options: The mongodb database connection parameters.
+    :type db_options: dict
+    :param base_path: Path to the top-level directory where to save files.
+    :type base_path: string
+    :return tuple The return code, the boot document id and errors.
+    """
+    ret_code = 201
+    doc_id = None
+    errors = {}
+
+    status = job_data['status']
+    if status not in LAVA_JOB_RESULT:
+        msg = "Invalid job status: {}".format(status)
+        ret_code = 404
+        utils.LOG.error(msg)
+        utils.errors.add_error(errors, ret_code, msg)
+        return ret_code, doc_id, errors
```

Comment:
Yes, I think I originally wanted to throw an exception here but then changed my mind as it's really validating the data sent from the client (LAVA).  I'll address the other issues first and see if this can be tidied up easily.